### PR TITLE
add test job to CI

### PR DESCRIPTION
**Pull request by Tayef-Shah**
### Notion ticket link

[Ensure automated tests pass CI](https://www.notion.so/uwblueprintexecs/Ensure-automated-tests-pass-in-CI-c883376b4076478790cbf97cd9269718)

### Implementation description

- Create a build job and a test job
- Run them sequentially, test job runs after build job passes
- Refer to [this template](https://circleci.com/docs/2.0/sample-config/#sequential-workflow) of a sequential workflow
- Add dependency caching, example [here](https://circleci.com/docs/2.0/yarn/#caching)
- Add workspace to save project state between jobs. Concept diagram [here](https://circleci.com/docs/2.0/concepts/#caches-workspaces-and-artifacts)
- Add environment variable under circleci project settings. [See here for reference](https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-project)

### What should reviewers focus on?

- Good CI practices

### Checklist

- [x] My PR name is descriptive and in imperative tense
- [x] I have run the linter
- [x] I have requested a review from the PL, and/or other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -6,12 +6,43 @@ jobs:
...
+      - image: circleci/python:3.9-node
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Unit Tests
+          command: yarn test . --passWithNoTests
+
+workflows:
```
Comment (jolenezheng) [inline .circleci/config.yml:42]: this is very cool

File: .circleci/config.yml [hunk 2]
```diff
@@ -6,12 +6,43 @@ jobs:
...
+  test:
+    docker:
+      - image: circleci/python:3.9-node
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Unit Tests
+          command: yarn test . --passWithNoTests
```
Comment (jolenezheng) [inline .circleci/config.yml:40]: just curious, are we taking out the passWithNoTests eventually?

**Review by jolenezheng — APPROVED** [review]
lgtm! this is really cool, didn't know circleci had all these configuration keys :3

**Review by jolenezheng — APPROVED** [review]
lgtm! this is really cool, didn't know circleci had all these configuration keys :3

File: .circleci/config.yml (hunk 2 above)
Comment (Tayef-Shah) [reply .circleci/config.yml:40]: yeah after we add tests lul

**Comment by Tayef-Shah** [comment]
> lgtm! this is really cool, didn't know circleci had all these configuration keys :3

I am happy to explain further if you want. For some reason I'm proud I brought the job time down from 2 mins to 1 min 20s

**Comment by jolenezheng** [comment]
> > lgtm! this is really cool, didn't know circleci had all these configuration keys :3
> 
> I am happy to explain further if you want. For some reason I'm proud I brought the job time down from 2 mins to 1 min 20s

one question, not sure if it makes any sense buut just curious - we're specifying the docker image twice one for the build and again for the tests. is there a way to just specify it once in the workflow or smth??

**Comment by Tayef-Shah** [comment]
> > > lgtm! this is really cool, didn't know circleci had all these configuration keys :3
> > 
> > 
> > I am happy to explain further if you want. For some reason I'm proud I brought the job time down from 2 mins to 1 min 20s
> 
> one question, not sure if it makes any sense buut just curious. we're specifying the docker image twice - one for the build and again for the tests. is there a way to just specify it once in the workflow or smth??
> 
> and yes that is proudworthy !!!!! 🌟 😌

Sadly, the Blueprint CircleCI plan does not offer [docker layer caching](https://circleci.com/docs/2.0/docker-layer-caching/) :( Another sad thing is that [saving docker images to a workspace](https://support.circleci.com/hc/en-us/articles/360019182513-Build-a-Docker-image-in-one-job-and-use-it-in-another-job) doesn't work well with images that are 1GB or larger (ours is 1.12GB). I think it should be fine tho, since it seems like the `docker` and `image` keys just defines the execetor/image (see [here](https://circleci.com/docs/2.0/executor-intro/)).

**Pull request merged**